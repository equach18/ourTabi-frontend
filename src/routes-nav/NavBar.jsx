import { NavLink } from "react-router-dom";
import { useContext } from "react";
import TripSearchBar from "../components/trip/TripSearchBar";
import UserContext from "../context/UserContext";

function NavBar() {
  const { currentUser, logout } = useContext(UserContext);

  return (
    <nav className="bg-stone-300 shadow px-6 py-4 flex items-center justify-between relative">
      <NavLink
        to={currentUser ? "/dashboard" : "/"}
        className="text-2xl font-bold text-orange-600 font-heading z-10 hover:text-orange-500 transition"
      >
        OurTabi
      </NavLink>

      {currentUser && (
        <div className="absolute left-1/2 transform -translate-x-1/2">
          <TripSearchBar />
        </div>
      )}

      <div className="flex items-center space-x-4 z-10">
        {currentUser ? (
          <button
            onClick={logout}
            className="text-slate-600 hover:text-orange-500 transition"
          >
            Logout
          </button>
        ) : (
          <>
            <NavLink
              to="/login"
              className="text-slate-600 hover:text-orange-500 transition"
            >
              Login
            </NavLink>
            <NavLink
              to="/signup"
              className="text-slate-600 hover:text-orange-500 transition"
            >
              Sign Up
            </NavLink>
          </>
        )}
      </div>
    </nav>
  );
}

export default NavBar;
