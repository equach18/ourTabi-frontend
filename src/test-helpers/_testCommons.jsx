import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import UserContext from "../context/UserContext";
import TripContext from "../context/TripContext";

export const testUser1 = {
  id: 1,
  firstName: "Test1",
  lastName: "User1",
  username: "testUser1",
  profilePic: "",
};
export const testUser2 = {
  id: 2,
  firstName: "Test2",
  lastName: "User2",
  username: "testUser2",
  profilePic: "testUser2.png",
};
export const testUser3 = {
  id: 3,
  firstName: "Test2",
  lastName: "User2",
  username: "testUser3",
  profilePic: "testUser3.png",
};

export const defaultMembers = [
  {
    userId: testUser1.id,
    username: testUser1.username,
    profilePic: testUser1.profilePic,
  },
  {
    userId: testUser2.id,
    username: testUser2.username,
    profilePic: testUser2.profilePic,
  },
];

export const testActivity = {
  id: 1,
  name: "Kayaking",
  tripId: 53,
  category: "Other",
  location: "Chicago",
  description: "So much fun",
  scheduledTime: "2025-07-10T14:30:00Z",
};

export const testComment = {
  id: 1,
  text: "Test comment",
  createdAt: "2025-07-11T14:30:00Z",
  userId: 1,
};

export const testTrip = {
  id: 42,
  title: "Test trip",
  destination: "Chicago, IL",
  startDate: "2025-07-10",
  endDate: "2025-07-12",
  isPrivate: true,
};

export const testFriendsData = {
  friends: [testUser1 ],
  incomingRequests: [testUser2],
  sentRequests: [testUser3],
};

// Default mock context values
const defaultUserContext = { currentUser: testUser1 };
const defaultTripContext = {
  voteOnActivity: vi.fn(),
  votes: {},
  deleteComment: vi.fn(),
  addMember: vi.fn(),
  removeMember: vi.fn(),
  members: defaultMembers,
};

// create mock trip context
export function createMockTripContext(overrides = {}) {
  return {
    voteOnActivity: vi.fn(),
    votes: {},
    deleteComment: vi.fn(),
    members: [],
    addMember: vi.fn(),
    removeMember: vi.fn(),
    ...overrides,
  };
}

/**
 * Renders a component inside MemoryRouter
 */
export function renderWithRouter(ui) {
  return render(<MemoryRouter>{ui}</MemoryRouter>);
}

/**
 * Renders a component inside MemoryRouter + UserContext + TripContext
 */
export function renderWithAllProviders(
  ui,
  {
    userContext = defaultUserContext,
    tripContext = defaultTripContext,
    route = "/",
  } = {}
) {
  return render(
    <MemoryRouter initialEntries={[route]}>
      <UserContext.Provider value={userContext}>
        <TripContext.Provider value={tripContext}>{ui}</TripContext.Provider>
      </UserContext.Provider>
    </MemoryRouter>
  );
}

/**
 * Renders a component with just UserContext
 */
export function renderWithUserProvider(ui, { userContext } = {}) {
  return render(
    <UserContext.Provider value={userContext}>{ui}</UserContext.Provider>
  );
}
