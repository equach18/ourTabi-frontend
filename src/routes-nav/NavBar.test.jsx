import { screen } from "@testing-library/react";
import NavBar from "./NavBar";
import { renderWithAllProviders } from "../test-helpers/_testCommons";

// mock TripSearchBar
vi.mock("../components/trip/TripSearchBar", () => ({
  default: () => <div data-testid="TripSearchBar">Mock SearchBar</div>,
}));

describe("NavBar", () => {
  function renderNavBar({ currentUser = null, logout = vi.fn() } = {}) {
    return renderWithAllProviders(<NavBar />, {
      userContext: { currentUser, logout, friends: [] }, // provide logout and friends just in case
      route: "/",
    });
  }

  it("matches snapshot when logged out", () => {
    const { asFragment } = renderNavBar();
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot when logged in", () => {
    const { asFragment } = renderNavBar({
      currentUser: { username: "test" },
      logout: vi.fn(),
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders login and signup links when logged out", () => {
    renderNavBar();

    expect(screen.getByText("Login")).toBeInTheDocument();
    expect(screen.getByText("Sign Up")).toBeInTheDocument();
    expect(screen.queryByText("Logout")).not.toBeInTheDocument();
    expect(screen.queryByTestId("TripSearchBar")).not.toBeInTheDocument();
  });

  it("renders logout and search bar when logged in", () => {
    renderNavBar({
      currentUser: { username: "test" },
      logout: vi.fn(),
    });

    expect(screen.getByText("Logout")).toBeInTheDocument();
    expect(screen.getByTestId("TripSearchBar")).toBeInTheDocument();
    expect(screen.queryByText("Login")).not.toBeInTheDocument();
  });

  it("calls logout function when logout button is clicked", () => {
    const mockLogout = vi.fn();

    renderNavBar({
      currentUser: { username: "test" },
      logout: mockLogout,
    });

    screen.getByText("Logout").click();
    expect(mockLogout).toHaveBeenCalled();
  });
});
