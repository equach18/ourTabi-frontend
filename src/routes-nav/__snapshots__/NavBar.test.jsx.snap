// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NavBar > matches snapshot when logged in 1`] = `
<DocumentFragment>
  <nav
    class="bg-stone-300 shadow px-6 py-4 flex items-center justify-between relative"
  >
    <a
      class="text-2xl font-bold text-orange-600 font-heading z-10 hover:text-orange-500 transition"
      data-discover="true"
      href="/dashboard"
    >
      OurTabi
    </a>
    <div
      class="absolute left-1/2 transform -translate-x-1/2"
    >
      <div
        data-testid="TripSearchBar"
      >
        Mock SearchBar
      </div>
    </div>
    <div
      class="flex items-center space-x-4 z-10"
    >
      <button
        class="text-slate-600 hover:text-orange-500 transition"
      >
        Logout
      </button>
    </div>
  </nav>
</DocumentFragment>
`;

exports[`NavBar > matches snapshot when logged out 1`] = `
<DocumentFragment>
  <nav
    class="bg-stone-300 shadow px-6 py-4 flex items-center justify-between relative"
  >
    <a
      aria-current="page"
      class="text-2xl font-bold text-orange-600 font-heading z-10 hover:text-orange-500 transition active"
      data-discover="true"
      href="/"
    >
      OurTabi
    </a>
    <div
      class="flex items-center space-x-4 z-10"
    >
      <a
        class="text-slate-600 hover:text-orange-500 transition"
        data-discover="true"
        href="/login"
      >
        Login
      </a>
      <a
        class="text-slate-600 hover:text-orange-500 transition"
        data-discover="true"
        href="/signup"
      >
        Sign Up
      </a>
    </div>
  </nav>
</DocumentFragment>
`;
